### PR TITLE
Fix resolve order

### PR DESCRIPTION
**Pull request by janderson2**
The use of go routines in dependency resolution was meaning that getpan was not processing the cpanfile in order as it claimed.  This was causing random failures due to dependencies of dependencies being resolved before our matching cpan dependencies that were higher in the file.  The resulting dependency conflict would cause a failure. 

Also fixed in the PR is the chance of a race condition during removal of install lock on a module in the global module map.

I have also go fmt'd and bumped the release version to 0.9.

### CODE REVIEW

File: getpan/getpan/dependency.go [hunk 1]
```diff
@@ -196,122 +198,13 @@ func (d *DependencyList) Resolve() error {
...
-				errorLock.Lock()
-				errs = append(errs, dep.Module.String())
-				errorLock.Unlock()
-				<-semaphore
-				return
-			}
-
-			<-semaphore
-		}(dep)
+		dep.Resolve(d.Parent)
```
Comment (emead) [inline getpan/getpan/dependency.go]: error return value not checked

File: getpan/getpan/dependency.go [hunk 2]
```diff
@@ -323,24 +216,91 @@ func (d *DependencyList) Resolve() error {
...
+			return errors.New(errstr)
+		}
+
+		log.Trace("Version %s matches %s", dep.Module, gm.Version)
+
+		// TODO See if downloading a new version would be better
+		dep.Module = gm
+	} else {
+		log.Debug("Downloading: %s", dep.Module)
+		err := dep.Module.Download()
```
Comment (eldeal) [inline getpan/getpan/dependency.go]: This error can be inlined with the conditional below

File: getpan/getpan/dependency.go [hunk 3]
```diff
@@ -323,24 +216,91 @@ func (d *DependencyList) Resolve() error {
...
+
+		if dep.Additional != nil && len(dep.Additional) > 0 {
+			log.Trace("Adding cpanfile additional REQS")
+			for _, additional := range dep.Additional {
+				log.Trace("Adding additional dependency from cpanfile: %s", additional)
+				dep.Module.Deps.AddDependency(additional)
+			}
+		}
+
+		err = dep.Module.loadDependencies()
```
Comment (eldeal) [inline getpan/getpan/dependency.go]: Inline also please

File: getpan/getpan/dependency.go [hunk 4]
```diff
@@ -323,24 +216,91 @@ func (d *DependencyList) Resolve() error {
 	return nil
 }
 
-func (v *Dependency) Resolve() error {
-	log.Trace("Resolving dependency: %s", v)
+func (dep *Dependency) Resolve(p *Module) error {
```
Comment (emead) [inline getpan/getpan/dependency.go]: Exported method should be preceded by a comment.

Comment (emead) [inline getpan/getpan/dependency.go]: `dep` should be renamed to `d` for consistency with other methods

File: getpan/getpan/dependency.go [hunk 5]
```diff
@@ -323,24 +216,91 @@ func (d *DependencyList) Resolve() error {
...
 		if m != nil {
 			log.Trace("=> Resolved dependency: %s", m)
-			v.Module = m
-			return nil
+			dep.Module = m
+			break
+		}
+	}
+	if dep.Module == nil {
+		log.Error("Error resolving dependency: %s", dep)
```
Comment (emead) [inline getpan/getpan/dependency.go]: dep is not a string, and this message implies an error occurred when it didn't.

Perhaps `log.Error("Dependency [%s] not resolved", d.Name)`

File: getpan/getpan/dependency.go [hunk 6]
```diff
@@ -323,24 +216,91 @@ func (d *DependencyList) Resolve() error {
...
 		if m != nil {
 			log.Trace("=> Resolved dependency: %s", m)
-			v.Module = m
-			return nil
+			d.Module = m
+			break
+		}
+	}
+	if d.Module == nil {
+		log.Error("Error resolving dependency: %s", d)
```
Comment (emead) [inline getpan/getpan/dependency.go]: d is not a string, and this message implies an error occurred when it didn't.

Perhaps `log.Error("Dependency [%s] not resolved", d.Name)`

File: getpan/getpan/dependency.go [hunk 7]
```diff
@@ -323,24 +216,91 @@ func (d *DependencyList) Resolve() error {
...
+		return errors.New(fmt.Sprintf("Dependency not found from any source: %s", d.String()))
```
Comment (emead) [inline getpan/getpan/dependency.go]: should replace `errors.New(fmt.Sprintf(...))` with `fmt.Errorf(...)`

File: getpan/getpan/dependency.go [hunk 8]
```diff
@@ -323,24 +216,91 @@ func (d *DependencyList) Resolve() error {
...
+		err := d.Module.Download()
+		if err != nil {
+			log.Error("Error downloading module %s: %s", d.Module, err)
+			return err
+		}
+
+		if p != nil {
+			if p.IsCircular(d.Module) {
+				log.Error("Detected circular dependency %s from module %s", d.Module, p)
+				return errors.New(fmt.Sprintf("Detected circular dependency %s from module %s", d.Module, p))
```
Comment (emead) [inline getpan/getpan/dependency.go]: should replace `errors.New(fmt.Sprintf(...))` with `fmt.Errorf(...)`

File: getpan/getpan/dependency.go (hunk 6 above)
Comment (janderson2) [inline getpan/getpan/dependency.go]: It can be used as a string since dependency implements stringer interface.

**Comment by eldeal** [comment]
:+1:

**Pull request merged**